### DISCUSSION TITLE
vitejs integration

### DISCUSSION THREAD

**Discussion by MadadiSaeed** [mswjs/msw#712] [Q&A]
I asked in vitejs github repository on how exclude code for production but unfortunately nobody answered my question. so I ask my question here so hopefully somebody maybe has answer to my question. The question is how to integrate msw with vitejs **?** How to exclude msw from production build **?**

**Comment by kettanaito** [comment]
Hey, @samadadi. 

There's a Vite example from @msutkowski as a part of https://github.com/mswjs/msw/issues/695:

- [Example repository](https://github.com/msutkowski/msw-ts-vitejs).

Looks like he's got a way to exclude the worker on production as well, if you follow the discussion in the said issue.

**Reply by MadadiSaeed** [reply]
Thank you for your respose. This example repository is not the solution at all. It is just a runtime check to use msw or not. At the end It includes msw in production build. The only way to exclude msw from production build is to write plugin to vite or rollup or webpack. If anyone wants, I can share my solution, although it is so simple.

**Reply by kettanaito** [reply]
Tools like webpack automatically eliminate "dead" code on a production build. The following expression is a dead code:

```js
if (false) {
  require('./mocks')
}
```

And should be stripped away by the bundler. I don't have enough intel to guarantee that, as bundlers change over time and this may no longer be the case. It may also require additional configuration for such dead code elimination to work.

**Reply by lukeknxt** [reply]
In case anyone is still dealing with Vite and MSW issues relating to the `public/` folder...

Whilst importing at runtime conditionally will exclude the library from the build, it won't exclude the `mockServiceWorker.js` file from being included in the build. To do that, you can use a plugin such as this in the Vite config (inspired by [this comment](https://github.com/vitejs/vite/discussions/7374#discussioncomment-2793892)).

```ts
// in vite.config.js

function excludeMsw() {
  return {
    name: "exclude-msw",
    resolveId(source) {
      return source === "virtual-module" ? source : null;
    },
    renderStart(outputOptions, _inputOptions) {
      const outDir = outputOptions.dir;
      const msWorker = path.resolve(outDir, "mockServiceWorker.js");
      fs.rm(msWorker, () => console.log(`Deleted ${msWorker}`));
    },
  };
}

export default defineConfig(({ mode }) => ({
  plugins: [react(), excludeMsw()],
  // rest of your config
})
```

**Reply by kingyue737** [reply]
@lukeknxt  Maybe you are looking for [vite-plugin-static-copy](https://github.com/sapphi-red/vite-plugin-static-copy). You can decide whether to copy `mockServiceWorker.js`  from `node_modules/mswjs` according to mode (dev or production).  With this plugin, you don't even need to put `mockServiceWorker.js` in `public` folder or commit it to your repo.

**Reply by velly** [reply]
> Hey, @samadadi.
> 
> There's a Vite example from @msutkowski as a part of #695:
> 
> * [Example repository](https://github.com/msutkowski/msw-ts-vitejs).
> 
> Looks like he's got a way to exclude the worker on production as well, if you follow the discussion in the said issue.

Why you placed the file mockServiceWorker.js under both public folder and src in demo?
I get Error: getaddrinfo ENOTFOUND mock at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:109:26)
Because the url/ mockServiceWorker.js for public is forbidden by Vite4. httpProxy errror.

any configuration to control the request url of mockServiceWorker.js?

**Comment by micha149** [comment]
You can split the mocking stuff into a separate module by using a dynamic import. The example @kettanaito posted, give us a good starting point. But instead of using a static import, we can import the msw script dynamically in the prepare function.

Here is a example from my codebase:
```tsx
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { RecoilRoot } from 'recoil';
import App from './App';

const prepare = async (): Promise<void> => {
    if (import.meta.env.DEV) {
        const { default: worker } = await import('./mock/browser');
        worker.start();
    }
};

// eslint-disable-next-line promise/catch-or-return, promise/always-return
prepare().then(() => {
    ReactDOM.render(
        <React.StrictMode>
            <RecoilRoot>
                <BrowserRouter>
                    <App />
                </BrowserRouter>
            </RecoilRoot>
        </React.StrictMode>,
        document.getElementById('root'),
    );
});
```

**Reply by kettanaito** [reply]
Would triggering a separate chunk for the mocks ensure that they are not included in the production build?

**Reply by micha149** [reply]
Not directly. But not including the separate chunk, wich ist done by the `process.env.NODE_ENV === 'development'` will eliminate the chunk entirely.

**Reply by micha149** [reply]
Okay… Now I wonder why the `process.env.NODE_ENV` is working. I replace it with `import.meta.env.DEV` in my answer, which is the official vite way.

**Reply by florianmatz** [reply]
@micha149 @kettanaito I managed doing it the following way:

```js
<script type="module">
      if (import.meta.env.DEV) {
          const { worker } = await import('/src/mocks/browser');
          worker.start();
      }
 </script>
```

in the index.html of my vite setup.

I had a problem to get it started with Playwright, where I wanted to have access to ```window.msw...``` defined in my browser.ts file. But Playwright always moaned, that window.msw was undefined. Probably, because the lazy import is too slow when done in React.

However, I've look for a way to be able to start my worker upon React and found the solution above.

**Reply by micha149** [reply]
@florianmatz The problem is, that your „other” module did not wait until the module `/src/mocks/browser` is loaded. So, you get a race condition wether your app or your api mock loads faster. You have to put your app initialization after the mock initialization somehow. When you look at my example, I do the `ReactDOM.render` in the `.then()` of the mock initialization. This way, I ensure that my app is initialized **after** the mock was loaded.

**Reply by florianmatz** [reply]
@micha149 Yap, thats true. I've changed it your way, but I have found out another working method:

browser.ts
```ts
import { SetupWorkerApi, setupWorker, rest } from 'msw';
import { handlers } from './handlers';

let wrk: SetupWorkerApi = {} as SetupWorkerApi;

if (import.meta.env.DEV) {
    // Configure a Service Worker with the given request handlers
    wrk = setupWorker(...handlers);

    // Store msw on window to access it in e2e-test
    window.msw = {
        worker: wrk,
        rest,
    };
}

export const worker = wrk;
```

That also avoids including it into the vite bundle. 
By putting the setup into the condition, vite obviously checks, that I'm doing nothing with the import in PROD and thus, throwing it out of the bundling. 

I've analyzed my bundle and checked the output size. It seems to work.

**Reply by micha149** [reply]
Ah, that's interesting. I will try this in my next project, too.

**Reply by kettanaito** [reply]
I know that @ryota-murakami has been doing some amazing job on MSW and Vite integrations, if I'm not mistaken. They can provide some feedback on this, perhaps.

**Reply by ryota-murakami** [reply]
@micha149 @kettanaito I'm sorry about super slow response. 🙇‍♂️
Vite devServer setting `process.env.NODE_ENV = 'development'` [at the](https://github.com/vitejs/vite/blob/adacaba1cf03256cf18a3494b4d7b053605d4a46/packages/vite/src/node/server/index.ts#L302) cli bootstrap.  
After that Vite [mapping values](https://github.com/vitejs/vite/blob/f05a81387b6901622c86c2833f42420b8e18e95e/packages/vite/src/node/plugins/define.ts#L14-L45) to `import.meta.env` based on `vite.config.ts`.  
So both ways are expecting working as same as `import.meta.env.DEV` and `process.env.NODE_ENV`.

In my project, working with `process.env.NODE_ENV` like this.

https://github.com/laststance/vite-react-ts-extended/blob/main/src/main.tsx

**Comment by velly** [comment]
I try to migration [mock-api ](https://github.com/mswjs)to Vue3 project, that supports browser mock and nodejs mock for testing services. 
It works pretty well for test case. but fail to migration for browser.

Before "Mock Service Worker" to intercepting the browser request. It tries to fetch mockServiceWorker.js from development server.

Error network req:
'http://localhost:5173/mockServiceWorker.js

I seem the httpproxy error from vite --debug
20:08:14 [vite] http proxy error at ServiceWorker.js:
Error: getaddrinfo ENOTFOUND mock
at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:109:26)
ncaught (in promise) Error: [MSW] Failed to register the Service Worker:

And mswjs report error as follow:

**Failed to register a ServiceWorker for scope ('http://localhost:5173/') with script ('http://localhost:5173/mockServiceWorker.js'): A bad HTTP response code (500) was received when fetching the script.
at getWorkerInstance (getWorkerInstance.ts:85:11)
at async startWorkerInstance (createStartHandler.ts:31:24)
at async SetupWorkerApi.start (setupWorker.ts:192:12)
getWorkerInstance @ getWorkerInstance.ts:85
await in getWorkerInstance
install @ mtk-core.ts:29
use @ runtime-core.esm-bundler.js:4381
@ main.ts:22

I think the problem is hosting static resource ???
I place the mockServiceWorker.js in public folder but Vite seems it doesn't support this kind of static file(file has dot?)
